Mathv6

dgrat opened the pull request:
Changes since version 5:
 - made the is_zero/is_equal type checks a bit more stringent
 - more use of std::* cmath under Linux

lucasdemarchi commented:
> made the is_zero/is_equal type checks a bit more stringent

do you have an example?

Do you plan to do any other change before this being considered for merging?

lucasdemarchi commented on libraries/AP_Math/AP_Math.h:
```diff
@@ -425,9 +425,26 @@ T usec_to_hz(const T &usec) {
+template <class T>
+T linear_interpolate(const T &low_output, const T &high_output,
+                     const T &var_value,
+                     const T &var_low, const T &var_high)
```
I have to say I'm not a fan of doing a template just for using templates, so I may be biased in the question:

What's the advantage of using a template here? Besides that, moving it to the header would mean the function may be duplicated in each compilation unit, no?

lucasdemarchi replied:
Checking with nm confirms that with your version the function is in fact inlined

lucasdemarchi commented on libraries/AP_Math/AP_Math.h:
```diff
@@ -17,153 +15,135 @@
-    return nsec / NSEC_PER_USEC;
+template<class T>
+T nsec_to_usec(const T &nsec) {
+    return nsec != 0 ? nsec / NSEC_PER_USEC : 0;
```
why check 0?

lucasdemarchi replied:
and why a reference?

dgrat replied:
I used a reference in case this function will ever be used for bigger types like vectors matrices.
The zero check is in case of integer types, which would create a program termination and because this check does not hurt. I guess, you still want the zero check out. We it then be pushed?

lucasdemarchi replied:
Em 14 de abr de 2016 5:33 AM, "Daniel Frenzel" <notifications@github.com>
escreveu:
>
> In libraries/AP_Math/AP_Math.h:
>
> >      return usec * NSEC_PER_USEC;
> >  }
> >
> > -inline uint32_t nsec_to_usec(uint32_t nsec)
> > -{
> > -    return nsec / NSEC_PER_USEC;
> > +template<class T>
> > +T nsec_to_usec(const T &nsec) {
> > +    return nsec != 0 ? nsec / NSEC_PER_USEC : 0;
>
> I used a reference in case this function will ever be used for bigger
types like vectors matrices.

I don't think it will ever be. If it is, we can change.

> The zero check is in case of integer types, which would create a program
termination and because this check does not hurt. I guess, you still want
the zero check out. We it then be pushed?

I was talking about this specific case. You are not dividing by it so
there's no div by 0 here.

Lucas De Marchi

lucasdemarchi commented on libraries/AP_Math/AP_Math.h:
```diff
@@ -17,153 +15,135 @@
-    return USEC_PER_SEC / freq;
+template<class T>
+T hz_to_usec(const T &freq) {
+    return freq != 0 ? USEC_PER_SEC / freq : 0;
```
same... no need for checking against 0.

lucasdemarchi commented on libraries/AP_Math/AP_Math.h:
```diff
@@ -2,13 +2,11 @@
 #include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
 #include <AP_Param/AP_Param.h>
-#include <cmath>
```
I'm a little bit confused of removing cmath from here

lucasdemarchi commented on libraries/AP_Math/AP_Math_Common.h:
```diff
@@ -0,0 +1,15 @@
+ * @brief: This magic functions swaps the value of the two input parameters
+ */
+template <class T> 
+void swap(T& a, T& b) {
```
In AP_Compass_HMC5843 we are using std::swap(). Is there any problem with that that we should prefer our own version?

dgrat replied:
Can be removed now. Was a leftover.

lucasdemarchi commented on libraries/AP_Math/AP_Math_Float.h:
```diff
@@ -0,0 +1,176 @@
+        static_assert(std::is_floating_point<T>::value, "ERROR - isfinite(): template parameter not of type float\n");
+        return (isnan(val) || isinf(val)) ? false : true;
+    }
+};
```
humn.. we should had already fixed this... just make sure to include cmath. Or is there anything wrong with it still?

dgrat replied:
This is from the time when it was necessary. I remove it.

dgrat replied:
No, it still leads to build error if I build for PX4. :D

lucasdemarchi commented:
@dgrat thanks for the reworks... I think most of the changes are fine. I think it would be better if I already bring them since we agree about them and I know it's a huge work for keeping them out of tree. What do you think?  A great part of the patch, for example is changing the pythagoras2() to norm(); constrain_* functions are fine; wrap_* functions probably fine, too. I added some comments above about things I'm still unsure. Thanks.

dgrat commented:
@lucasdemarchi 
If you want I can remove the zero checks and check whether the swap function is still necessary. 
I will update this PR one more time maybe.

khancyr commented on libraries/AP_Math/AP_Math.h:
```diff
@@ -425,9 +425,26 @@ T usec_to_hz(const T &usec) {
+    }
+    // avoid zero divisions or zero like divisions
+    auto var_diff = var_high - var_low;
+    if(is_zero((float)var_diff)) {
```
it shouldn't be is_zero<float> ?

khancyr commented on libraries/AP_Math/tests/test_math.cpp:
```diff
@@ -0,0 +1,309 @@
+TEST(ZeroTest, zero)
+{
+  #define TEST_IS_ZERO(_test, _result) { \
+    bool bZero = is_zero(_test); \
```
it shouldn't be is_zero<float> ?

khancyr replied:
is_zero should works only on float and ints. 
So we should test int too.
And test failure on double maybe ?

dgrat replied:
No the compiler can figure this out himself. 
Only if people mix types you need to help sometimes.

khancyr commented on libraries/AP_Math/tests/test_math.cpp:
```diff
@@ -0,0 +1,309 @@
+  
+  TEST_IS_EQUAL(0.1,     0.10001,           false);
+  TEST_IS_EQUAL(0.1,     -0.1001,           false);
+  TEST_IS_EQUAL(0.f,     0.0f,              true);
```
We should pass on non zero value and negative value too
TEST_IS_EQUAL(10.f,     10.f,              true);
TEST_IS_EQUAL(-10.f,     -10.f,              true);

And int 
TEST_IS_EQUAL(10,    10,              true);
TEST_IS_EQUAL(-10,   -10,              true);
TEST_IS_EQUAL(10,    11,              false);
TEST_IS_EQUAL(10,    -10,              false);

khancyr commented on libraries/AP_Math/tests/test_math.cpp:
```diff
@@ -0,0 +1,309 @@
+  EXPECT_EQ(sq_4, 0.f);
+  EXPECT_EQ(sq_5, 8.f);
+  EXPECT_EQ(sq_6, 16.f);
+}
```
miss negative value
float sq_7 = sq(2,-2); 
EXPECT_EQ(sq_7, 8.f);

dgrat commented:
@lucasdemarchi 
I removed the commit with the zero checks and removed the swap function.
The other function cannot be removed because it is not present for PX4 users.
cmath is included by AP_Math_Float/Double. 

@khancyr 
I can try to improve gtest further later. At the moment there are more tests than before :D

khancyr commented:
mmokay github ...just remove my previous comment....
I was saying too that tests can be improved later .

The only mistake I see is the use of is_zero in linear_interpolate template instead of is_zero<float>  no ?

dgrat commented:
@khancyr 
You don't need to help the compiler here.

khancyr commented:
Ok! I was'nt totally sure on that

dgrat commented:
@lucasdemarchi 

> made the is_zero/is_equal type checks a bit more stringent
> do you have an example?

static_assert(std::is_arithmetic<FloatOne>::value, "ERROR - is_equal(): template parameter not of type float or int\n");

dgrat commented:
@lucasdemarchi 
From my site, it is ready to be pushed :D

lucasdemarchi commented:
@dgrat I started the careful review to apply the patches. One thing that I noticed is that the compiler is always going to inline the template function. So the more we call the function the more the size of the binary increases.  I isolated the changes to wrap_* functions as example. Just changing them and nothing else increases the binary size in 3k!

I think the problem is not the increased size, but why it's increasing... which I think we need to workaround.  Is there a way not to have them inlined and still use the template?

lucasdemarchi commented:
Here is the branch with the isolated change: https://github.com/lucasdemarchi/ardupilot/commits/review/pr-3897

lucasdemarchi commented:
> If you want I can remove the zero checks and check whether the swap function is still necessary.

I was mentioning specifically in the cases where the 0 check was innocuous in which there were no div by 0 actually.

lucasdemarchi commented:
If I move the functions to the .cpp and instantiate the functions explicitly it gets down from 3k to 1.5k in size increased. I'm not sure if I like but I think it's a good compromise, no? At lease we know the size won't increase when we add calls to the functions. @dgrat what do you think?
https://github.com/lucasdemarchi/ardupilot/commits/review/pr-3897-v2

lucasdemarchi commented:
size comparisons:

```
master:
 908252	   2820	  57612	 968684	  ec7ec	/home/lucas/p/dronecode/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf

template wrap_*
 911228	   2820	  57612	 971660	  ed38c	/home/lucas/p/dronecode/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf

template wrap_* with explicitly instantiation
 909740	   2820	  57612	 970172	  ecdbc	/home/lucas/p/dronecode/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf
```

dgrat commented:
@lucasdemarchi 
Thanks Lucas, typical compiler madness. I will insert this into my commits.

dgrat commented:
I made your suggestion a bit more cxx11. 

```
make[1]: Leaving directory '/home/dgrat/ardupilot'
   text    data     bss     dec     hex filename
 987212    2828   57676 1047716   ffca4 /home/dgrat/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf
PX4 ArduCopter Firmware is in ArduCopter-v2.px4

make[1]: Leaving directory '/home/dgrat/ardupilot'
   text    data     bss     dec     hex filename
 985900    2828   57676 1046404   ff784 /home/dgrat/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf
```

dgrat commented:
@lucasdemarchi 
At the very end I am surprised about the fact that the size increase is so small.
Of course not for the DBL_MATH version, because of the std::functions, but for the other platforms.

OXINARF commented:
@dgrat You might want to resend the last commit because Github didn't recognized your username - it seems it was commited with a wrong e-mail?

dgrat commented:
@OXINARF 
thanks
